testutils/testutils.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2018 The Prometheus Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testutils
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+// GatherAndCompare retrieves all metrics exposed by a collector and compares it
+// to an expected output in the Prometheus text exposition format.
+// metricNames allows only comparing the given metrics. All are compared if it's nil.
+func GatherAndCompare(c prometheus.Collector, expected string, metricNames ...string) error {
+	expected = removeUnusedWhitespace(expected)
+
+	reg := prometheus.NewPedanticRegistry()
+	if err := reg.Register(c); err != nil {
```

Comment:
How about leaving these two lines to the caller? I.e. make `GatherAndCompare` accept a `Gatherer` rather than a `Collector`. This would also let the name of the function appear more consistent.

With this approach, a caller could act on the registry level, which allows testing even if metrics are coming from a library that does not allow access to individual collectors, but a registry with all the collectors already registered is readily available. Also, the caller can decide if a pedantic registry should be used or not.

If a caller wants to test a single Collector, it's easy to wrap `GatherAndCompare` with these two lines. Or wo could even provide a `CollectAndCompare`.

Reply:
As a concrete example, right now you couldn't test many Collectors because the "cheat" by providing a dummy Desc, which the pedantic registry would flag. (#47 will allow to implement those Collectors cleanly, and I'll work in this issue this evening (coincidentally).)

Reply:
:+1: